pkg/mimir/modules.go

```diff
@@ -684,6 +684,10 @@ func (t *Mimir) initQueryFrontendTripperware() (serv services.Service, err error
 	t.QueryFrontendCodec = querymiddleware.NewPrometheusCodec(t.Registerer, t.Cfg.Frontend.QueryMiddleware.QueryResultResponseFormat)
 	promqlEngineRegisterer := prometheus.WrapRegistererWith(prometheus.Labels{"engine": "query-frontend"}, t.Registerer)
 
+	// This indirection is necessary because the query-frontend service depends on the query-frontend tripperware service, so
+	// doesn't exist at the point where this method is called.
+	getQueryFrontendService := func() services.Service { return t.ServiceMap[QueryFrontend] }
```

Comment:
did you consider doing this in initQueryFrontend? Perhaps we can fit it in the handler or warp the tripperware there with the frontendRunningRoundTripper.

Reply:
Very good point, I think I was focusing too much on the idea that this should be middleware, but now that it's a RoundTripper, then doing it in `initQueryFrontend` is straightforward. 

Fixed in https://github.com/grafana/mimir/pull/6621/commits/f3fae2757248af36ef04c9788508f543137a8db2